wolfcrypt/src/port/Espressif/esp_crt_bundle/esp_crt_bundle.c

```diff
@@ -0,0 +1,1380 @@
+ *   for reference:
+ *     typedef int (*VerifyCallback)(int, WOLFSSL_X509_STORE_CTX*);
+ *
+ * This is the callback for TLS handshake verify / valiation. See related:
+ *   wolfssl_ssl_conf_verify_cb_no_signer
+ *   wolfssl_ssl_conf_verify_cb_before_date
+ *   wolfssl_ssl_conf_verify_cb_after_date
+ *
+ * This callback is called FOR EACH cert in the store.
+ * Not all certs in the store will have a match for a cert in the bundle,
+ * but we NEED ONE to match when a preverify error occurs.
+ *
+ * See wolfssl_ssl_conf_verify() for setting callback to this function.
+ * Typically set when calling esp_crt_bundle_attach(). Specifically:
+ *    cfg->crt_bundle_attach(&tls->conf) in esp_tls_wolfssl.c
+ *    from the ESP-IDF esp-tls component.
+ *
+ * See esp_tls.h file: esp_err_t (*crt_bundle_attach)(void *conf)
+ *   and initialization in esp_transport_ssl_crt_bundle_attach
+ *       from the tcp_transport component: (transport_ssl.c)
+ *
+ * Functions in esp_crt_bundle are same names as other providers and
+ * gated in as appropriate when enabling CONFIG_ESP_TLS_USING_WOLFSSL.
+ *
+ * Note the wolfSSL component CMakeLists.txt *MUST* be properly linked in the
+ * file to be used within the ESP-IDF. Something like this:
+ *
+ *   target_link_libraries(${COMPONENT_LIB} PUBLIC ${wolfssl})
+ *
+ * Returns:
+ * 0 if the verification process should stop immediately with an error.
+ * 1 if the verification process should continue with the rest of handshake. */
+static CB_INLINE int wolfssl_ssl_conf_verify_cb(int preverify,
+                                      WOLFSSL_X509_STORE_CTX* store)
+{
+#ifdef WOLFSSL_DEBUG_CERT_BUNDLE
+    char before_str[32];
+    char after_str[32];
+    WOLFSSL_ASN1_TIME *notBefore = NULL;
+    WOLFSSL_ASN1_TIME *notAfter = NULL;
+
+    int initial_preverify;
+    initial_preverify = preverify;
+    /* Show the interesting preverify & error state upon entry to callback. */
+    if (preverify == 1) {
+        ESP_LOGCBI(TAG, "preverify == 1\n");
+    }
+    else {
+        ESP_LOGCBW(TAG, "preverify == %d\n", preverify);
+    }
+
+    if (store->error == 0) {
+        ESP_LOGCBI(TAG, "store->error == 0");
+    }
+    else {
+        ESP_LOGCBW(TAG, "store->error: %d", store->error);
+    }
+
+
+    notBefore = wolfSSL_X509_get_notBefore(store->current_cert);
+    wolfSSL_ASN1_TIME_to_string(notBefore, before_str, sizeof(before_str));
+    ESP_LOGCBI(TAG, "Not Before: %s", before_str);
+
+    esp_show_current_datetime();
+
+    notAfter = wolfSSL_X509_get_notAfter(store->current_cert);
+    wolfSSL_ASN1_TIME_to_string(notAfter, after_str, sizeof(after_str));
+    ESP_LOGCBI(TAG, "Not After: %s", after_str);
+#endif
+
+    /* One possible condition is the error "Failed to find signer".
+     * This is where we search the bundle for a matching needed CA cert. */
+    if ((preverify == 0) && (store->error == ASN_NO_SIGNER_E)) {
+        ESP_LOGCBW(TAG, "Setting _need_bundle_cert!");
+        _need_bundle_cert = 1;
```

Comment:
Also see where this is set, but not used.

Reply:
Same. This is intended for end users. Added `wolfssl_need_bundle_cert()` to file: `esp_crt_bundle.h` .